super-ddf.c

```diff
@@ -877,30 +877,170 @@ static void *load_section(int fd, struct ddf_super *super, void *buf,
 	return buf;
 }
 
-static int load_ddf_headers(int fd, struct ddf_super *super, char *devname)
+
+/* Search for DDF_HEADER_MAGIC in the last 32MB of the device */
+static int search_for_ddf_headers(int fd, char *devname,
+				  unsigned long long *out)
 {
 	unsigned long long dsize;
+	int result = 0;
 
+	/* Get device size */
 	get_dev_size(fd, NULL, &dsize);
 
-	if (lseek64(fd, dsize - 512, 0) == -1L) {
-		if (devname)
-			pr_err("Cannot seek to anchor block on %s: %s\n",
+	const size_t READ_BLOCK_SIZE = 4096;
+	const unsigned long long SEARCH_REGION_SIZE = 32 * 1024 * 1024;
+
+	/* Determine the search range */
+	unsigned long long search_start =
+	    (dsize > SEARCH_REGION_SIZE) ? (dsize - SEARCH_REGION_SIZE) : 0;
+	unsigned long long search_end = dsize;
+	unsigned long long pos = search_start;
+
+	void *buffer = NULL;
+	be32 *magic_ptr = NULL;
+
+	/* Allocate aligned memory for reading */
+	if (posix_memalign(&buffer, READ_BLOCK_SIZE, READ_BLOCK_SIZE) != 0) {
+		perror("posix_memalign");
```

Comment:
Thank you for identifying that. I have submitted a new function named `xmemalign` in `xmalloc.c to handle posix_memalign, which exits if the allocation fails.

However, there are many direct function calls in the `ddf` code, so it will take some time to transition to using this new function.